### ISSUE TITLE
Multi-threaded/multi-process serving?

### ISSUE THREAD

**Issue by rory** [mojodna/tessera#41]
I've started looking into tessera and it seems really cool. As someone with no nodejs experience, I like how I can just call it on the command line without having to programme things in nodejs.

However, from what I can see, tessera only runs single threaded/single process. It doesn't appear to be able to handle more than one incoming request at a time. I found this out by running apache bench (`ab`) against a tessera server with 100 concurrent requests, and only saw one CPU-worth of processing in my `htop`.

Is it possible to run tessera in a multi-threaded/multi-process way?

**Comment by mojodna** [comment]
JavaScript code is inherently single-threaded, however Node includes some internal threadpools for handling filesystem operations and facilitating multi-threaded native code. Mapnik is one of the bindings that takes advantage of the threadpools, so if you were rendering new tiles on each request, you'd see multiple cores in use.

Theoretically, tessera could use [`cluster`](https://nodejs.org/docs/latest/api/cluster.html) to run multiple processes, but that would add confusion (and increased non-CPU resource usage) when combined with modules that take advantage of the threadpool. In other words, it would be somewhat beneficial for static files, but at that point you'd be better off using a different web server, as tessera is intended for the creation of new tiles.

I suspect that the reason you're only seeing a single CPU-worth of processing is that `ab` is repeatedly requesting the same URL, which will be internally cached (in single-threaded JavaScript land) after the first request.

**Comment by mojodna** [comment]
tl;dr: no, by design, because the CPU-intensive tasks that tessera invokes (e.g. mapnik) are multi-threaded (by using Node's internal threadpool). To tweak the size of the threadpool, set `UV_THREADPOOL_SIZE`. tessera sets what seems to be a reasonable default (https://github.com/mojodna/tessera/blob/master/server.js#L5) if one wasn't provided, though that's worth tweaking if Mapnik is i/o-limited.
